lib/extras.py

```diff
@@ -401,9 +401,15 @@ def initialize(self, logobj):
         self._logobj = logobj
         if _logging and isinstance(
                 logobj, (_logging.Logger, _logging.LoggerAdapter)):
-            self.log = self._logtologger
+            self._log = _weakref.WeakMethod(self._logtologger)
         else:
-            self.log = self._logtofile
+            self._log = _weakref.WeakMethod(self._logtofile)
+
+    def log(self, *args, **kwargs):
+        """
+        Public interface of the log method defined in initialize
+        """
+        return self._log()(*args, **kwargs)
```

Comment:
You should check if `self._log()` returns None. In that case I think you can just plonk the message: it likely means that the program is being terminated.

Reply:
So you mean something like this?
`

    if self._log:
         return self._log()(*args, **kwargs)

`

Reply:
No, more:

```python
log = self._log()
if log:
    log(*args, **kwargs)
```

Reply:
I'm not a 100% sure of this myself but wouldn't this replicate the same bound method issue again?

It is possible that the WeaktMethod does decrease the reference count even after we do this though.

Reply:
> I'm not a 100% sure of this myself but wouldn't this replicate the same bound method issue again?

No, because this would be into the `log()` function. `log`, here, is a local variable.

Reply:
I also wanted to ask if this issue might be happening in other parts. I could make the same changes there too.